src/alternative-gas-token/L1ScrollMessengerNonETH.sol

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.24;
+
+import {AddressUpgradeable} from "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol";
+
+import {IL1MessageQueue} from "../L1/rollup/IL1MessageQueue.sol";
+import {IScrollChain} from "../L1/rollup/IScrollChain.sol";
+import {L1ScrollMessenger} from "../L1/L1ScrollMessenger.sol";
+import {IMessageDropCallback} from "../libraries/callbacks/IMessageDropCallback.sol";
+import {ScrollConstants} from "../libraries/constants/ScrollConstants.sol";
+import {WithdrawTrieVerifier} from "../libraries/verifier/WithdrawTrieVerifier.sol";
+
+contract L1ScrollMessengerNonETH is L1ScrollMessenger {
+    /**********
+     * Errors *
+     **********/
+
+    error ErrorDuplicatedMessage();
+
+    error ErrorNonZeroValueFromCaller();
+
+    error ErrorInsufficientMsgValue();
+
+    /*************
+     * Constants *
+     *************/
+
+    /// @notice The address of `L1NativeTokenGateway` contract.
+    address public immutable nativeTokenGateway;
+
+    /***************
+     * Constructor *
+     ***************/
+
+    constructor(
+        address _nativeTokenGateway,
+        address _counterpart,
+        address _rollup,
+        address _messageQueue
+    ) L1ScrollMessenger(_counterpart, _rollup, _messageQueue) {
+        nativeTokenGateway = _nativeTokenGateway;
+    }
+
+    /**********************
+     * Internal Functions *
+     **********************/
+
+    /// @inheritdoc L1ScrollMessenger
+    function _sendMessage(
+        address _to,
+        uint256 _value,
+        bytes memory _message,
+        uint256 _gasLimit,
+        address _refundAddress
+    ) internal override {
+        // if we want to pass value to L2, must call from `L1NativeTokenGateway`.
+        if (_value > 0 && _msgSender() != nativeTokenGateway) revert ErrorNonZeroValueFromCaller();
```

Comment:
Will it be possible to use the batch bridge gateway with `nativeTokenGateway`?

Reply:
no, need to modify the `L1BatchBridgeGateway` and `L2BatchBridgeGateway` to support.